gateway/api/views/files.py

```diff
@@ -67,38 +69,164 @@ def check_user_has_provider(self, user, provider_name):
+            funtion (Program): the Qiskit Function that is going to be checked
+
+        Returns:
+            bool: boolean value that verifies if the user has access or not to the function
+        """
+
+        instances = function.instances.all()
+        user_groups = user.groups.all()
+        has_access = any(group in instances for group in user_groups)
+        if not has_access:
+            logger.error(
+                "User [%s] has no access to function [%s/%s].",
+                user.id,
+                function.provider.name,
+                function.title,
+            )
+        return has_access
+
     def list(self, request):
-        """List of available for user files."""
-        response = Response(
-            {"message": "Requested file was not found."},
-            status=status.HTTP_404_NOT_FOUND,
-        )
-        files = []
+        """
+        It returns a list with the names of available files for the user directory:
+            it will look under its username or username/provider_name/function_title
+        """
+
         tracer = trace.get_tracer("gateway.tracer")
         ctx = TraceContextTextMapPropagator().extract(carrier=request.headers)
         with tracer.start_as_current_span("gateway.files.list", context=ctx):
-            user_dir = request.user.username
-            provider_name = request.query_params.get("provider")
-            if provider_name is not None:
-                if self.check_user_has_provider(request.user, provider_name):
-                    user_dir = provider_name
-                else:
-                    return response
-            user_dir = os.path.join(
-                sanitize_file_path(settings.MEDIA_ROOT),
-                sanitize_file_path(user_dir),
+            username = request.user.username
+            provider_name = sanitize_name(request.query_params.get("provider"))
+            function_title = sanitize_name(request.query_params.get("function"))
+            working_dir = USER_STORAGE
+
+            function = None
+            if provider_name:
+                function = self.get_provider_function(
+                    provider_name=provider_name, function_title=function_title
+                )
+                if not function:
+                    return Response(
+                        {
+                            "message": f"Qiskit Function {provider_name}/{function_title} doesn't exist."  # pylint: disable=line-too-long
+                        },
+                        status=status.HTTP_404_NOT_FOUND,
+                    )
+
+            if function:
+                if not self.user_has_access_to_provider_function(
+                    request.user, function
+                ):
+                    return Response(
+                        {"message": "You don't have access to this Qiskit Function."},
+                        status=status.HTTP_403_FORBIDDEN,
+                    )
```

Comment:
Well it has sense because functions can be: "provider_name/function_title" or just "function_title" but definitely I have a bug there because the second function is always `None`. I will review this logic, thanks Goyo!

Reply:
Awesome catch, I love you ❤️ 

This comment is fixed in [fe9b501](https://github.com/Qiskit/qiskit-serverless/pull/1529/commits/fe9b50165090aff71b9151c5780a9d20a425d4d5)

Basically the `if` was not tabulated inside that if as you mentioned and I added an additional check for user functions.